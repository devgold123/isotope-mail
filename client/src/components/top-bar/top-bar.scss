@import '../../styles/partials/variables';
@import '@material/top-app-bar/mdc-top-app-bar';

.topBar {
  @extend .mdc-top-app-bar;
  margin-left: $drawerWidth;
  width: calc(100% - #{$drawerWidth});
}
