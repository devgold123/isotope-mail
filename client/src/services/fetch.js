export const HttpHeaders = {
  ISOTOPE_CREDENTIALS: 'X-Isotope-Credentials',
  ISOTOPE_SALT: 'X-Isotope-Salt',
  CONTENT_TYPE: 'Content-Type'
};

export const HttpStatusFamilies = {
  INFORMATIONAL: 1,
  SUCCESSFUL: 2,
  REDIRECTION: 3,
  CLIENT_ERROR: 4,
  SERVER_ERROR: 5
};

/**
 * Object to store the different AbortController(s) that will be used in the service methods to fetch from the API backend.
 *
 * @type {Object}
 * @private
 */
export const abortControllerWrappers = {};

/**
 * Converts response Promise to json Promise or throws an error.
 *
 * @param {Response}response
 * @returns {PromiseLike<object>}
 */
export function toJson(response) {
  if (!response.ok) {
    throw Error(`${response.status}`);
  }
  return response.json();
}

/**
 * Triggers the abort method of a the specified AbortController
 *
 * @param abortController {AbortController}
 */
export function abortFetch(abortController) {
  if (abortController && abortController.abort) {
    abortController.abort();
  }
}

/**
 * Creates a new headers object including Isotope specific credentials headers for server authentication.
 *
 * @param {{}} credentials
 * @param {{}} [originalHeaders={}] originalHeaders
 * @returns {{}}
 */
export function credentialsHeaders(credentials, originalHeaders) {
  const newHeaders = Object.assign({}, originalHeaders ? originalHeaders : {});
  newHeaders[HttpHeaders.ISOTOPE_CREDENTIALS] = credentials.encrypted;
  newHeaders[HttpHeaders.ISOTOPE_SALT] = credentials.salt;
  return newHeaders;
}

/**
 * Returns true if the response the Successful HTTP status code range (2xx)
 *
 * @param HTTP status code of the response
 * @returns {boolean} true if status code belongs to HttpStatusFamilies.SUCCESSFUL family
 */
export function isSuccessful(status) {
  return Math.floor(status / 100) === HttpStatusFamilies.SUCCESSFUL;
}
