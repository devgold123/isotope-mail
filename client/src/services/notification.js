const GRANTED_PERMISSION = 'granted';
const DENIED_PERMISSION = 'denied';

function requestPermission() {
  // Request only if necessary
  if ([GRANTED_PERMISSION, DENIED_PERMISSION].includes(Notification.permission)) {
    return Promise.resolve(Notification.permission);
  }
  return Notification.requestPermission();
}

export function notify(message, options) {
  return requestPermission().then(permission => {
    if (permission === GRANTED_PERMISSION) {
      const notification = new Notification(message, {...options, icon: 'assets/images/notification-icon.png'});
      notification.onclick = () => window.focus();
      return notification;
    }
    return null;
  });
}
