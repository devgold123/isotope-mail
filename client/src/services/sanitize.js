import sanitize from 'dompurify';
import ReactDOMServer from 'react-dom/server';
import {renderSpinner} from '../components/spinner/spinner';

sanitize.addHook('afterSanitizeAttributes', node => {
  // set all elements owning target to target=_blank
  if ('target' in node) {
    node.setAttribute('target', '_blank');
  }
  // set non-HTML/MathML links to xlink:show=new
  if (!node.hasAttribute('target')
    && (node.hasAttribute('xlink:href')
      || node.hasAttribute('href'))) {
    node.setAttribute('xlink:show', 'new');
  }
});

const ISOTOPE_WHITELISTED_URL = '#IsotopeEmbedded';

// 1/2 Replaces "insecure" blob: uris before sanitation for controlled ones
sanitize.addHook('uponSanitizeAttribute', (node, hookEvent) => {
  if (node.nodeName === 'IMG' && hookEvent.attrName === 'src' && hookEvent.attrValue.indexOf('blob:') === 0) {
    node.src = `${node.src.substring(5)}${ISOTOPE_WHITELISTED_URL}`;
    hookEvent.attrValue = node.src;
  }
});

// 2/2 Replaces previously parsed "insecure" urls to original ones -> Whitelist effect
// Hides cid: images that are being loaded
sanitize.addHook('afterSanitizeAttributes', node => {
  if (node.nodeName === 'IMG' && node.src && node.src.indexOf(ISOTOPE_WHITELISTED_URL) > -1) {
    node.src = `blob:${node.src.replace(ISOTOPE_WHITELISTED_URL, '')}`;
  }
  if (node.nodeName === 'IMG' && node.src && node.src.indexOf('cid:') === 0) {
    const spinner = document.createElement('div');
    spinner.innerHTML = ReactDOMServer.renderToStaticMarkup(renderSpinner());
    spinner.getElementsByClassName('canvas')[0].style.height = node.height ? node.height : '22px';
    node.parentElement.replaceChild(spinner, node);
  }
});
export default sanitize;
