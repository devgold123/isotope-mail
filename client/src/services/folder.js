import {URLS} from './url';
import {
  backendRequest as applicationBackendRequest,
  backendRequestCompleted as applicationBackendRequestCompleted,
  renameFolder as renameFolderAction} from '../actions/application';
import {backendRequest, setFolders, updateFolder} from '../actions/folders';
import {abortControllerWrappers, abortFetch, credentialsHeaders, toJson} from './fetch';

export const FolderTypes = Object.freeze({
  INBOX: {serverName: 'INBOX', icon: 'inbox', position: 0},
  DRAFTS: {attribute: '\\Drafts', icon: 'drafts', position: 1},
  SENT: {attribute: '\\Sent', icon: 'send', position: 2},
  TRASH: {attribute: '\\Trash', icon: 'delete', position: 3},
  FOLDER: {icon: 'folder'}
});

/**
 * Processes an initial folder list and adds the corresponding {@link FolderTypes}
 *
 * @param initialFolders {Array}
 * @returns {Array}
 */
export function processFolders(initialFolders) {
  if (!initialFolders) {
    return null;
  }
  const folders = [];
  const specialFolders = [];
  initialFolders.map(folder => {
    if (folder.name && folder.name.toUpperCase() === FolderTypes.INBOX.serverName) {
      folder.type = FolderTypes.INBOX;
      specialFolders[folder.type.position] = folder;
    } else {
      folder.type = FolderTypes.FOLDER;
      // Identify special folder
      let special = false;
      for (const t of [FolderTypes.DRAFTS, FolderTypes.SENT, FolderTypes.TRASH]) {
        if (folder.attributes && folder.attributes.includes(t.attribute)) {
          folder.type = t;
          specialFolders[folder.type.position] = folder;
          special = true;
          break;
        }
      }
      // If regular folder just add to the folder array
      if (!special) {
        folders.push(folder);
      }
    }
    folder.children = processFolders(folder.children);
  });
  // Insert special folders in specific positions
  specialFolders.reverse().forEach(f => folders.unshift(f));
  return folders;
}

/**
 * Builds an object containing all of the folders and their children at the same level
 */
export function explodeFolders(originalFolders, explodedFolders = {}) {
  originalFolders.forEach(f => {
    explodedFolders[f.folderId] = f;
    if (f.children && f.children.length > 0) {
      explodeFolders(f.children, explodedFolders);
    }
  });
  return explodedFolders;
}

export async function getFolders(dispatch, credentials, loadChildren) {
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  abortControllerWrappers.getFoldersAbortController = new AbortController();
  const signal = abortControllerWrappers.getFoldersAbortController.signal;

  const url = new URL(URLS.FOLDERS, window.location.origin);
  if (loadChildren) {
    url.search = new URLSearchParams({loadChildren: true}).toString();
  }
  dispatch(backendRequest());
  const response = await fetch(url, {
    method: 'GET',
    headers: credentialsHeaders(credentials),
    signal: signal
  });
  const folders = await toJson(response);
  dispatch(setFolders(folders));
}

export function renameFolder(dispatch, credentials, folderToRename, newName) {
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  dispatch(applicationBackendRequest());
  fetch(folderToRename._links.rename.href, {
    method: 'PUT',
    headers: credentialsHeaders(credentials, {'Content-Type': 'application/json'}),
    body: newName
  })
    .then(toJson)
    .then(renamedFolderParent => {
      if (renamedFolderParent.fullName.length === 0) {
        // Root folder (replace current folder tree)
        dispatch(setFolders(renamedFolderParent.children));
      } else {
        dispatch(updateFolder(renamedFolderParent));
      }
      dispatch(renameFolderAction(null));
      dispatch(applicationBackendRequestCompleted());
    })
    .catch(error => {
      dispatch(applicationBackendRequestCompleted());
    });
}

