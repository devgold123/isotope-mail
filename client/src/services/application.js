import sjcl from 'sjcl';
import {URLS} from './url';
import {
  backendRequest,
  backendRequestCompleted,
  editMessage,
  selectFolder,
  setUserCredentials
} from '../actions/application';
import {toJson} from './fetch';
import {FolderTypes, getFolders} from './folder';
import i18n from './i18n';
import {recoverState} from './indexed-db';
import {setFolders} from '../actions/folders';
import {setCache} from '../actions/messages';
import {resetFolderMessagesCache} from './message';
import sanitize from './sanitize';

export const DEFAULT_IMAP_PORT = 993;
export const DEFAULT_IMAP_SSL = true;
export const DEFAULT_SMTP_PORT = 465;
export const DEFAULT_SMTP_SSL = true;

/**
 * @typedef {Object} Credentials
 * @property {string} serverHost - IMAP host
 * @property {string} serverPort - IMAP port
 * @property {string} user - User
 * @property {string} password - Password
 */
/**
 * Performs a login to the IMAP server using the API.
 *
 * If the server validates the credentials, previous state is reloaded (if exists) from IndexedDB.
 *
 * If previous state exists, message cache, folder list, and selected folder will be pre-populated with existing values.
 * An initial call to the backend will be made to refresh the list of messages of the selected folder.
 *
 * If a previous state doesn't exist, an initial load of the first level folders will be requested to the server.
 *
 * @param dispatch {Dispatch & function}
 * @param credentials {Credentials}
 * @returns {Promise<void>}
 */
export async function login(dispatch, credentials) {
  dispatch(backendRequest());
  const url = URLS.LOGIN;
  // Will be used as the key in the IndexedDB
  const userId = sjcl.codec.hex.fromBits(
    sjcl.hash.sha256.hash(`${credentials.serverHost}|${credentials.user}`));
  // Will be used as the encryption password to store state in the IndexedDB
  const hash = sjcl.codec.hex.fromBits(
    sjcl.hash.sha256.hash(`${credentials.serverHost}|${credentials.user}|${credentials.password}`));

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(credentials)
  });
  dispatch(backendRequestCompleted());
  if (response.ok) {
    const validatedCredentials = await toJson(response);
    dispatch(setUserCredentials(userId, hash, validatedCredentials));
    // Reload data from indexedDb
    const recoveredState = await recoverState(userId, hash);
    if (recoveredState !== null) {
      await dispatch(setCache(recoveredState.messages.cache));
      await dispatch(setFolders(recoveredState.folders.items));
      dispatch(selectFolder(recoveredState.application.selectedFolder));
    } else {
      // Retrieve first level folders to show something ASAP
      const folders = await getFolders(dispatch, validatedCredentials, false);
      // Retrieve and select INBOX folder so that user has something in the screen
      const inbox = folders.find(f => f.type === FolderTypes.INBOX);
      if (inbox) {
        dispatch(selectFolder(inbox));
        resetFolderMessagesCache(dispatch, validatedCredentials, inbox, null);
      }
    }
  }
}

export function editNewMessage(dispatch) {
  dispatch(editMessage({to: [], cc: [], bcc: [], subject: '', content: ''}));
}

export function replyMessage(dispatch, originalMessage) {
  const recipients = [...originalMessage.recipients];
  const recipientMapper = r => r.address;
  const to = recipients.filter(r => r.type === 'To').map(recipientMapper).concat(originalMessage.from);
  const cc = recipients.filter(r => r.type === 'Cc').map(recipientMapper);
  const bcc = recipients.filter(r => r.type === 'Bcc').map(recipientMapper);
  const subject = `Re: ${originalMessage.subject}`;
  const formattedDate = new Date(originalMessage.receivedDate).toLocaleString(navigator.language, {
    year: 'numeric', month: '2-digit', day: '2-digit', hour: '2-digit', minute: '2-digit', second: '2-digit'
  });

  const t = i18n.t.bind(i18n);
  const content = `
    <p></p>
    <hr/>
    <p>
      <b>${t('replyAction.From')}:</b> ${originalMessage.from.join(', ')}<br/>
      <b>${t('replyAction.Date')}:</b> ${formattedDate}<br/>
      <b>${t('replyAction.Subject')}:</b> ${originalMessage.subject}<br/>
    </p>
    <br/>
    ${sanitize.sanitize(originalMessage.content)}
  `;

  dispatch(editMessage({to, cc, bcc, subject, content
  }));
}
