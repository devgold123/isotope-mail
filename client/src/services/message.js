import {
  backendRequest as aBackendRequest,
  backendRequestCompleted as aBackendRequestCompleted, replaceMessageEmbeddedImages
} from '../actions/application';
import {
  backendRequest,
  backendRequestCompleted,
  deleteFromCache,
  setFolderCache,
  updateCache
} from '../actions/messages';
import {refreshMessage} from '../actions/application';
import {updateFolder} from '../actions/folders';
import {abortControllerWrappers, abortFetch, credentialsHeaders, toJson} from './fetch';
import {processFolders} from './folder';
import {persistMessageCache} from './indexed-db';
import {KEY_HASH, KEY_USER_ID} from './state';

const _eventSourceWrappers = {};

function _closeEventSource(dispatch, es) {
  if (es && es.close) {
    es.close();
    if (!es.DISPATCHED) {
      dispatch(backendRequestCompleted());
      es.DISPATCHED = true;
      es.resolvePromise();
    }
  }
}

function _readContent(dispatch, credentials, folder, message, signal, attachment) {
  fetch(attachment._links.download.href, {
    method: 'GET',
    headers: credentialsHeaders(credentials),
    signal: signal
  })
    .then(response => response.blob())
    .then(blob => {
      dispatch(replaceMessageEmbeddedImages(folder, message, attachment, blob));
    });
}

/**
 *
 * @param dispatch
 * @param credentials
 * @param folder
 */
export async function resetFolderMessagesCache(dispatch, credentials, folder) {
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  if (folder && folder._links) {
    const allMessages = [];
    // Prefer EventSourcePolyfill instead of EventSource to allow sending HTTP headers in all browsers
    const es = new window.EventSourcePolyfill(folder._links.messages.href,
      {
        headers: credentialsHeaders(credentials)
      });
    _eventSourceWrappers.resetFolderMessagesCache = es;
    dispatch(backendRequest());
    es.onmessage = e => {
      const messages = JSON.parse(e.data);
      allMessages.push(...messages);
      if (e.lastEventId === '1') {
        // This is the last batch
        dispatch(setFolderCache(folder, allMessages));
        _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
        // Manually persist newest version of message cache
        persistMessageCache(
          sessionStorage.getItem(KEY_USER_ID), sessionStorage.getItem(KEY_HASH), folder, [...allMessages]);
      } else {
        dispatch(updateCache(folder, messages));
      }
    };
    // Return a promise
    const eventSourceCompleted = new Promise(resolve => {
      // Store resolve function that will be called in _closeEventSource
      es.resolvePromise = resolve;
    });
    return eventSourceCompleted;
  }
  return null;
}

/**
 *
 * @param dispatch
 * @param credentials
 * @param folder
 * @param message {object}
 */
export function readMessage(dispatch, credentials, folder, message) {
  // Abort + signal
  if (message && message._links) {
    abortFetch(abortControllerWrappers.readMessageAbortController);
    abortControllerWrappers.readMessageAbortController = new AbortController();
    const signal = abortControllerWrappers.readMessageAbortController.signal;

    dispatch(aBackendRequest());
    fetch(message._links.self.href, {
      method: 'GET',
      headers: credentialsHeaders(credentials),
      signal: signal
    })
      .then(response => {
        dispatch(aBackendRequestCompleted());
        return response;
      })
      .then(toJson)
      .then(completeMessage => {
        dispatch(refreshMessage(folder, completeMessage));
        // Update folder with freshest information
        dispatch(updateFolder(processFolders([completeMessage.folder])[0]));
        // Update folder cache with message marked as read (don't store content in cache)
        const messageWithNoContent = {...completeMessage};
        delete messageWithNoContent.content;
        dispatch(updateCache(folder, [messageWithNoContent]));
        // Read message's embedded images
        completeMessage.attachments
          .filter(a => a.contentId && a.contentId.length > 0)
          .filter(a => a.contentType.toLowerCase().indexOf('image/') === 0)
          .forEach(a => _readContent(dispatch, credentials, folder, message, signal, a));
      }).catch(() => dispatch(aBackendRequestCompleted()));
  }
}

export function downloadAttachment(credentials, attachment) {
  const fetch$ = fetch(attachment._links.download.href, {
    method: 'GET',
    headers: credentialsHeaders(credentials)
  }).then(response => response.blob());
  fetch$
    .then(blob => {
      if (navigator.msSaveBlob) {
        navigator.msSaveBlob(blob, attachment.fileName);
      } else {
        const url = URL.createObjectURL(blob);
        const tempLink = document.createElement('a');
        tempLink.href = url;
        tempLink.download = attachment.fileName;
        document.body.appendChild(tempLink);
        // tempLink.click();
        tempLink.dispatchEvent(new MouseEvent('click', {bubbles: true, cancelable: true, view: window}));
        document.body.removeChild(tempLink);
        URL.revokeObjectURL(url);
      }
    });
  return fetch$;
}

export function moveMessage(dispatch, credentials, fromFolder, toFolder, message) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  // Calculate new fromFolder values
  const fromFolderUpdated = {
    ...fromFolder,
    messageCount: fromFolder.messageCount - 1,
    unreadMessageCount: fromFolder.unreadMessageCount - (message.seen ? 0 : 1),
    newMessageCount: fromFolder.newMessageCount - (message.recent ? 1 : 0)
  };

  // Update state with expected response from server
  dispatch(deleteFromCache(fromFolder, [message]));
  dispatch(updateFolder(fromFolderUpdated));
  fetch(message._links.move.href.replace('{toFolderId}', toFolder.folderId), {
    method: 'PUT',
    headers: credentialsHeaders(credentials)
  })
    .then(toJson)
    .then(newMessages => {
      if (Array.isArray(newMessages)) {
        dispatch(updateCache(toFolder, newMessages));
        newMessages.forEach(m => dispatch(updateFolder(processFolders([m.folder])[0])));
      }
    })
    .catch(() => {
      // Rollback state from dispatched expected responses
      dispatch(updateCache(fromFolder, [message]));
      dispatch(updateFolder(fromFolder));
    });
}

export function setMessageSeen(dispatch, credentials, folder, message, seen) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  // Update state with expected response from server
  const expectedUpdatedMessage = {...message, seen: seen};
  dispatch(updateCache(folder, [expectedUpdatedMessage]));
  const expectedUpdatedFolder = {...folder, unreadMessageCount: folder.unreadMessageCount + 1};
  dispatch(updateFolder(expectedUpdatedFolder));

  fetch(message._links.seen.href, {
    method: 'PUT',
    headers: credentialsHeaders(credentials, { 'Content-Type': 'application/json'}),
    body: JSON.stringify(seen)
  })
    .then(toJson)
    .then(updatedMessage => {
      dispatch(updateCache(folder, [updatedMessage]));
      dispatch(updateFolder(processFolders([updatedMessage.folder])[0]));
    })
    .catch(() => {
      // Rollback state from dispatched expected responses
      dispatch(updateCache(folder, [message]));
      dispatch(updateFolder(folder));
    });
}
