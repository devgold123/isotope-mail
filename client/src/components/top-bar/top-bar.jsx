import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {FolderTypes} from '../../services/folder';
import {selectMessage} from '../../actions/application';
import {moveMessages, setMessageSeen} from '../../services/message';
import styles from './top-bar.scss';
import mainCss from '../../styles/main.scss';

class TopBar extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const collapsed = this.props.sideBarCollapsed;
    const isMessageViewer = this.props.selectedMessage && Object.keys(this.props.selectedMessage).length > 0;
    let title = this.props.title;
    if (this.props.selectedFolder && this.props.selectedFolder.name
      && this.props.selectedFolder.type !== FolderTypes.INBOX) {
      title = `${this.props.selectedFolder.name} - ${title}`;
    }
    return (
      <header className={`${styles.topBar} ${styles['with-custom-styles']}
      ${collapsed ? '' : styles['with-side-bar']}
      ${mainCss['mdc-top-app-bar']} ${mainCss['mdc-top-app-bar--fixed']}`}>
        <div className={mainCss['mdc-top-app-bar__row']}>
          <section className={`${mainCss['mdc-top-app-bar__section']} ${mainCss['mdc-top-app-bar__section--align-start']}`}>
            {collapsed ?
              <button onClick={this.props.sideBarToggle}
                className={`material-icons ${mainCss['mdc-top-app-bar__navigation-icon']}`}>
                menu
              </button> :
              null
            }
            {isMessageViewer ?
              <Fragment>
                <button onClick={() => this.props.selectMessage(null)}
                  className={`material-icons ${mainCss['mdc-top-app-bar__navigation-icon']}`}>
                  arrow_back
                </button>
              </Fragment>
              :
              <span className={mainCss['mdc-top-app-bar__title']}>{title}</span>
            }
          </section>
          <section className={`${mainCss['mdc-top-app-bar__section']} ${mainCss['mdc-top-app-bar__section--align-end']}`}>
            {isMessageViewer ?
              <Fragment>
                <button
                  onClick={this.props.deleteMessage}
                  className={`material-icons ${mainCss['mdc-top-app-bar__action-item']}`}>
                  delete
                </button>
                <button
                  onClick={this.props.toggleMessageSeen}
                  className={`material-icons ${mainCss['mdc-top-app-bar__action-item']}`}>
                  markunread
                </button>
              </Fragment>
              :
              null
            }
          </section>
        </div>
      </header>
    );
  }
}

TopBar.propTypes = {
  title: PropTypes.string.isRequired,
  selectedFolder: PropTypes.object,
  selectMessage: PropTypes.func.isRequired,
  sideBarToggle: PropTypes.func.isRequired,
  sideBarCollapsed: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
  title: state.application.title,
  selectedFolder: state.folders.explodedItems[state.application.selectedFolderId] || null,
  selectedMessage: state.application.selectedMessage,
  credentials: state.application.user.credentials,
  folders: state.folders
});

const mapDispatchToProps = dispatch => ({
  selectMessage: message => dispatch(selectMessage(message)),
  deleteMessage: (credentials, folders, selectedFolder, selectedMessage) => {
    let trashFolder = Object.values(folders.explodedItems).find(f => f.type === FolderTypes.TRASH);
    if (!trashFolder) {
      trashFolder = folders.items.find(f => f.name.toUpperCase() === 'TRASH');
    }
    if (selectedMessage && selectedFolder && trashFolder) {
      moveMessages(dispatch, credentials, selectedFolder, trashFolder, [selectedMessage]);
      dispatch(selectMessage(null));
    }
  },
  toggleMessageSeen: (credentials, selectedFolder, selectedMessage) => {
    setMessageSeen(dispatch, credentials, selectedFolder, selectedMessage, !selectedMessage.seen);
    dispatch(selectMessage(null));
  }
});

const mergeProps = (stateProps, dispatchProps, ownProps) => (Object.assign({}, stateProps, dispatchProps, ownProps, {
  deleteMessage: () =>
    dispatchProps.deleteMessage(
      stateProps.credentials, stateProps.folders, stateProps.selectedFolder, stateProps.selectedMessage),
  toggleMessageSeen: () =>
    dispatchProps.toggleMessageSeen(
      stateProps.credentials, stateProps.selectedFolder, stateProps.selectedMessage)
}));

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(TopBar);
