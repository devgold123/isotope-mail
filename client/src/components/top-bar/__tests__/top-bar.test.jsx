import React from 'react';
import {shallow} from 'enzyme/build/index';
import {TopBar} from '../top-bar';
import {FolderTypes} from '../../../services/folder';

describe('TopBar component test suite', () => {
  describe('Snapshot render', () => {
    test('TopBar with selected folder and no new message and no selected message, should render in list mode', () => {
      // Given
      const props = {
        sideBarCollapsed: false,
        sideBarToggle: jest.fn(),
        selectedFolder: {type: FolderTypes.FOLDER, name: 'SelectedFolder'},
        selectedMessages: [],
        title: 'Isotope Mail Client - TITLE'};

      // When
      const topBar = shallow(<TopBar {...props}/>);

      // Then
      expect(topBar).toMatchSnapshot();
    });
    test('TopBar with selected folder (and messageIds) and no new message and no selected message, should render in list mode with delete and mark read buttons', () => {
      // Given
      const props = {
        sideBarCollapsed: false,
        sideBarToggle: jest.fn(),
        selectedFolder: {type: FolderTypes.FOLDER, name: 'SelectedFolder'},
        selectedMessages: [{uid: 1}],
        title: 'Isotope Mail Client - TITLE'};

      // When
      const topBar = shallow(<TopBar {...props}/>);

      // Then
      expect(topBar).toMatchSnapshot();
    });
    test('TopBar with selected message and no new message and no selected folder, should render in view mode with buttons', () => {
      // Given
      const props = {
        sideBarCollapsed: false,
        sideBarToggle: jest.fn(),
        selectedMessage: {uid: 1},
        title: 'Isotope Mail Client - TITLE',
        outbox: null};

      // When
      const topBar = shallow(<TopBar {...props}/>);

      // Then
      expect(topBar).toMatchSnapshot();
    });
    test('TopBar with new message and no selected message and no selected folder, should render in edit mode without buttons', () => {
      // Given
      const props = {
        sideBarCollapsed: false,
        sideBarToggle: jest.fn(),
        newMessage: {subject: 'New Message'},
        title: 'Isotope Mail Client - TITLE'};

      // When
      const topBar = shallow(<TopBar {...props}/>);

      // Then
      expect(topBar).toMatchSnapshot();
    });
  });
});
