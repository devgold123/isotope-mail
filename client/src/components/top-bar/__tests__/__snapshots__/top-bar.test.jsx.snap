// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar component test suite Snapshot render TopBar with new message and no selected message and no selected folder, should render in edit mode without buttons 1`] = `
<header
  className="topBar with-custom-styles
      with-side-bar
      mdc-top-app-bar mdc-top-app-bar--fixed"
>
  <div
    className="mdc-top-app-bar__row"
  >
    <section
      className="mdc-top-app-bar__section mdc-top-app-bar__section--align-start"
    >
      <span
        className="mdc-top-app-bar__title"
      >
        Isotope Mail Client - TITLE
      </span>
    </section>
  </div>
  <Translate(ConfirmDeleteFromTrashDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</header>
`;

exports[`TopBar component test suite Snapshot render TopBar with selected folder (and messageIds) and no new message and no selected message, should render in list mode with delete and mark read buttons 1`] = `
<header
  className="topBar with-custom-styles
      with-side-bar
      mdc-top-app-bar mdc-top-app-bar--fixed"
>
  <div
    className="mdc-top-app-bar__row"
  >
    <section
      className="mdc-top-app-bar__section mdc-top-app-bar__section--align-start"
    >
      <span
        className="mdc-top-app-bar__title"
      >
        SelectedFolder - Isotope Mail Client - TITLE
      </span>
    </section>
    <section
      className="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"
    >
      <button
        className="material-icons mdc-top-app-bar__action-item"
        onClick={[Function]}
      >
        delete
      </button>
      <button
        className="material-icons mdc-top-app-bar__action-item"
        onClick={[Function]}
      >
        markunread
      </button>
    </section>
  </div>
  <Translate(ConfirmDeleteFromTrashDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</header>
`;

exports[`TopBar component test suite Snapshot render TopBar with selected folder and no new message and no selected message, should render in list mode 1`] = `
<header
  className="topBar with-custom-styles
      with-side-bar
      mdc-top-app-bar mdc-top-app-bar--fixed"
>
  <div
    className="mdc-top-app-bar__row"
  >
    <section
      className="mdc-top-app-bar__section mdc-top-app-bar__section--align-start"
    >
      <span
        className="mdc-top-app-bar__title"
      >
        SelectedFolder - Isotope Mail Client - TITLE
      </span>
    </section>
    <section
      className="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"
    />
  </div>
  <Translate(ConfirmDeleteFromTrashDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</header>
`;

exports[`TopBar component test suite Snapshot render TopBar with selected message and no new message and no selected folder, should render in view mode with buttons 1`] = `
<header
  className="topBar with-custom-styles
      with-side-bar
      mdc-top-app-bar mdc-top-app-bar--fixed"
>
  <div
    className="mdc-top-app-bar__row"
  >
    <section
      className="mdc-top-app-bar__section mdc-top-app-bar__section--align-start"
    >
      <button
        className="material-icons mdc-top-app-bar__navigation-icon"
        onClick={[Function]}
      >
        arrow_back
      </button>
    </section>
    <section
      className="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"
    >
      <button
        className="material-icons mdc-top-app-bar__action-item"
      >
        reply_all
      </button>
      <button
        className="material-icons mdc-top-app-bar__action-item"
      >
        forward
      </button>
      <button
        className="material-icons mdc-top-app-bar__action-item"
        onClick={[Function]}
      >
        delete
      </button>
      <button
        className="material-icons mdc-top-app-bar__action-item"
      >
        markunread
      </button>
    </section>
  </div>
  <Translate(ConfirmDeleteFromTrashDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</header>
`;
